File: sandpile_2x2_full_alphabet_search.py
#!/usr/bin/env python3
"""Exact full-alphabet search over every stable 2x2 background on Z^2.

Inputs a,b in {0,1,2,3} add a*p and b*p grains at the two top cells.
The diagonally paired bottom-cell odometer parities must equal a,b.
Sparse stabilization makes the computation genuinely infinite-lattice.
"""

from __future__ import annotations

import argparse
from collections import defaultdict, deque
from itertools import product


A = (0, 0)
B = (0, 1)
D = (1, 0)
C = (1, 1)
DIRECTIONS = ((-1, 0), (1, 0), (0, -1), (0, 1))


def stabilize(
    core: tuple[int, int, int, int],
    a_grains: int,
    b_grains: int,
) -> defaultdict[tuple[int, int], int]:
    state: defaultdict[tuple[int, int], int] = defaultdict(int)
    odometer: defaultdict[tuple[int, int], int] = defaultdict(int)
    for site, value in zip((A, B, D, C), core, strict=True):
        state[site] = value
    state[A] += a_grains
    state[B] += b_grains
    pending = deque(site for site in (A, B) if state[site] >= 4)
    queued = set(pending)
    while pending:
        site = pending.popleft()
        queued.discard(site)
        amount = state[site] // 4
        if amount == 0:
            continue
        state[site] -= 4 * amount
        odometer[site] += amount
        row, column = site
        for dr, dc in DIRECTIONS:
            neighbor = (row + dr, column + dc)
            state[neighbor] += amount
            if state[neighbor] >= 4 and neighbor not in queued:
                pending.append(neighbor)
                queued.add(neighbor)
    return odometer


def outputs(
    core: tuple[int, int, int, int], a: int, b: int, pulse: int
) -> tuple[int, int]:
    odometer = stabilize(core, a * pulse, b * pulse)
    return odometer[C], odometer[D]


def boolean_valid(
    core: tuple[int, int, int, int], pulse: int
) -> bool:
    return all(
        tuple(value & 1 for value in outputs(core, a, b, pulse))
        == (a, b)
        for a, b in ((1, 0), (0, 1), (1, 1))
    )


def full_valid(
    core: tuple[int, int, int, int], pulse: int
) -> tuple[bool, tuple[tuple[tuple[int, int], ...], ...]]:
    return full_valid_scales(core, pulse, pulse)


def full_valid_scales(
    core: tuple[int, int, int, int], a_pulse: int, b_pulse: int
) -> tuple[bool, tuple[tuple[tuple[int, int], ...], ...]]:
    table = tuple(
        tuple(
            (
                lambda odometer: (odometer[C], odometer[D])
            )(stabilize(core, a * a_pulse, b * b_pulse))
            for b in range(4)
        )
        for a in range(4)
    )
    valid = all(
        tuple(value & 1 for value in table[a][b])
        == (a & 1, b & 1)
        for a in range(4)
        for b in range(4)
    )
    return valid, table


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--maximum-pulse", type=int, default=500)
    args = parser.parse_args()
    boolean_hits = 0
    best_errors = 33
    best = None
    for pulse in range(1, args.maximum_pulse + 1):
        pulse_hits = 0
        for core in product(range(4), repeat=4):
            if not boolean_valid(core, pulse):
                continue
            boolean_hits += 1
            pulse_hits += 1
            valid, table = full_valid(core, pulse)
            errors = sum(
                ((table[a][b][0] & 1) != (a & 1))
                + ((table[a][b][1] & 1) != (b & 1))
                for a in range(4)
                for b in range(4)
            )
            if errors < best_errors:
                best_errors = errors
                best = pulse, core, table
                print(
                    f"best errors={errors} pulse={pulse} core={core}",
                    flush=True,
                )
            if valid:
                print(
                    f"FULL HIT pulse={pulse} core={core} table={table}",
                    flush=True,
                )
                return
        if pulse_hits:
            print(
                f"pulse={pulse} boolean_hits={pulse_hits} "
                f"cumulative={boolean_hits}",
                flush=True,
            )
    print(
        f"NO FULL HIT pulses=1..{args.maximum_pulse}; "
        f"boolean_hits={boolean_hits}; best_errors={best_errors}; best={best}"
    )


if __name__ == "__main__":
    main()

File: sandpile_2x2_full_alphabet_search.cpp
#include <array>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <queue>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

using Site = std::int64_t;

static Site pack(int row, int column) {
  return (static_cast<Site>(row) << 32) ^
         static_cast<std::uint32_t>(column);
}

static std::pair<int, int> unpack(Site site) {
  return {static_cast<int>(site >> 32),
          static_cast<int>(static_cast<std::uint32_t>(site))};
}

constexpr int A = 0;
constexpr int B = 1;
constexpr int D = 2;
constexpr int C = 3;

class Pile {
 public:
  explicit Pile(const std::array<int, 4>& core) {
    state_[pack(0, 0)] = core[A];
    state_[pack(0, 1)] = core[B];
    state_[pack(1, 0)] = core[D];
    state_[pack(1, 1)] = core[C];
  }

  void add(int a_grains, int b_grains) {
    std::queue<Site> queue;
    std::unordered_map<Site, bool> queued;
    auto seed = [&](Site site, int amount) {
      state_[site] += amount;
      if (state_[site] >= 4 && !queued[site]) {
        queued[site] = true;
        queue.push(site);
      }
    };
    if (a_grains) seed(pack(0, 0), a_grains);
    if (b_grains) seed(pack(0, 1), b_grains);
    while (!queue.empty()) {
      const Site site = queue.front();
      queue.pop();
      queued[site] = false;
      const int amount = state_[site] / 4;
      if (!amount) continue;
      state_[site] -= 4 * amount;
      odometer_[site] += amount;
      const auto [row, column] = unpack(site);
      for (const Site neighbor :
           {pack(row - 1, column), pack(row + 1, column),
            pack(row, column - 1), pack(row, column + 1)}) {
        state_[neighbor] += amount;
        if (state_[neighbor] >= 4 && !queued[neighbor]) {
          queued[neighbor] = true;
          queue.push(neighbor);
        }
      }
    }
  }

  std::array<int, 2> outputs() const {
    return {get(odometer_, pack(1, 1)),
            get(odometer_, pack(1, 0))};
  }

 private:
  static int get(const std::unordered_map<Site, int>& values, Site site) {
    const auto found = values.find(site);
    return found == values.end() ? 0 : found->second;
  }

  std::unordered_map<Site, int> state_;
  std::unordered_map<Site, int> odometer_;
};

static std::array<int, 2> run(const std::array<int, 4>& core,
                              int a_grains, int b_grains) {
  Pile pile(core);
  pile.add(a_grains, b_grains);
  return pile.outputs();
}

int main(int argc, char** argv) {
  int maximum_pulse = 10000;
  if (argc == 3 && std::string(argv[1]) == "--maximum-pulse") {
    maximum_pulse = std::stoi(argv[2]);
  } else if (argc != 1) {
    std::cerr << "usage: " << argv[0] << " [--maximum-pulse N]\n";
    return 2;
  }
  int best_errors = 33;
  std::array<int, 4> best_core{};
  int best_pulse = 0;
  std::array<std::array<std::array<int, 2>, 4>, 4> best_table{};
  std::uint64_t boolean_hits = 0;

  for (int code = 0; code < 256; ++code) {
    int value = code;
    std::array<int, 4> core{};
    for (int q = 3; q >= 0; --q) {
      core[q] = value % 4;
      value /= 4;
    }
    Pile north(core);
    Pile west(core);
    Pile both(core);
    for (int pulse = 1; pulse <= maximum_pulse; ++pulse) {
      north.add(1, 0);
      west.add(0, 1);
      both.add(1, 1);
      const auto n = north.outputs();
      const auto w = west.outputs();
      const auto nw = both.outputs();
      if ((n[0] & 1) != 1 || (n[1] & 1) != 0 ||
          (w[0] & 1) != 0 || (w[1] & 1) != 1 ||
          (nw[0] & 1) != 1 || (nw[1] & 1) != 1) {
        continue;
      }
      ++boolean_hits;
      int errors = 0;
      std::array<std::array<std::array<int, 2>, 4>, 4> table{};
      for (int a = 0; a < 4; ++a) {
        for (int b = 0; b < 4; ++b) {
          table[a][b] = run(core, a * pulse, b * pulse);
          errors += ((table[a][b][0] & 1) != (a & 1));
          errors += ((table[a][b][1] & 1) != (b & 1));
        }
      }
      if (errors < best_errors) {
        best_errors = errors;
        best_core = core;
        best_pulse = pulse;
        best_table = table;
        std::cout << "best errors=" << errors << " pulse=" << pulse
                  << " core=(" << core[0] << "," << core[1] << ","
                  << core[2] << "," << core[3] << ")\n";
        std::cout.flush();
      }
      if (errors == 0) {
        std::cout << "FULL HIT\n";
        goto finished;
      }
    }
    if ((code + 1) % 16 == 0) {
      std::cout << "cores=" << code + 1 << "/256 boolean_hits="
                << boolean_hits << " best_errors=" << best_errors << "\n";
      std::cout.flush();
    }
  }

finished:
  std::cout << "DONE max_pulse=" << maximum_pulse
            << " boolean_hits=" << boolean_hits
            << " best_errors=" << best_errors
            << " best_pulse=" << best_pulse << " best_core=("
            << best_core[0] << "," << best_core[1] << ","
            << best_core[2] << "," << best_core[3] << ")\n";
  for (int a = 0; a < 4; ++a) {
    std::cout << "a=" << a << ":";
    for (int b = 0; b < 4; ++b) {
      std::cout << " (" << best_table[a][b][0] << ","
                << best_table[a][b][1] << ")";
    }
    std::cout << "\n";
  }
}
